robotiq_force_torque_sensor/nodes/rq_sensor.cpp

```diff
@@ -40,99 +41,224 @@
 #include <stdio.h>
+#include <string.h>
 
 #include "boost/bind.hpp"
-#include "ros/ros.h"
-#include "std_msgs/String.h"
-#include "std_msgs/Bool.h"
-#include "std_msgs/Empty.h"
 #include "geometry_msgs/WrenchStamped.h"
-#include "robotiq_force_torque_sensor/rq_sensor_state.h"
 #include "robotiq_force_torque_sensor/ft_sensor.h"
+#include "robotiq_force_torque_sensor/rq_sensor_state.h"
 #include "robotiq_force_torque_sensor/sensor_accessor.h"
+#include "ros/ros.h"
+#include "std_msgs/Bool.h"
+#include "std_msgs/Empty.h"
+#include "std_msgs/String.h"
+
+class SensorBase
+{
+public:
+    SensorBase() : _connection_pub(_nh.advertise<std_msgs::Bool>("robotiq_force_torque_sensor_connected", 1))
+    {
+        ros::param::param<std::string>("~frame_id", _wrench_msg.header.frame_id, "robotiq_force_torque_frame_id");
+    }
 
-static void decode_message_and_do(INT_8 const  * const buff, INT_8 * const ret);
-bool wait_for_other_connection(void);
-bool try_connection(void);
-static int max_retries_(5);
-void shutdown(void);
-void simulateSensor(ros::NodeHandle& nh);
-void launchRealSensor(ros::NodeHandle& nh);
-ros::Publisher sensor_pub;
-ros::Publisher wrench_pub;
-ros::Publisher watchdog_pub;
-ros::ServiceServer service;
+    virtual ~SensorBase() = 0;
+    virtual bool makeConnection() = 0;
+
+    void Connect();
+    bool isConnected() const
+    {
+        return _connection_msg.data;
+    }
+
+    void publishSensor();
+    void publishWatchdog()
+    {
+        _watchdog_pub.publish(std_msgs::Empty());
+    }
+
+    void publishConnection()
+    {
+        _connection_pub.publish(_connection_msg);
+        ros::spinOnce();
+    }
+
+protected:
+    void setConnection(bool newState)
+    {
+        _connection_msg.data = newState;
+        publishConnection();
+    }
+
+private:
+    virtual bool _getSensorData(robotiq_force_torque_sensor::ft_sensor &msgStream) = 0;
+    bool _receiverCallback(robotiq_force_torque_sensor::sensor_accessor::Request &req, robotiq_force_torque_sensor::sensor_accessor::Response &res);
+    virtual void _decodeMessageAndDo(INT_8 const *const buff, INT_8 *const ret) = 0;
+
+    ros::NodeHandle _nh;
+    std_msgs::Bool _connection_msg;
+    ros::Publisher _connection_pub;
+
+    robotiq_force_torque_sensor::ft_sensor msgStream;
+    ros::Publisher _sensor_pub;
+
+    geometry_msgs::WrenchStamped _wrench_msg;
+    ros::Publisher _wrench_pub;
+    ros::Publisher _watchdog_pub;
+    ros::ServiceServer _service;
+};
+
+class SimulatedSensor : public SensorBase
+{
+public:
+    bool makeConnection() override
+    {
+        setConnection(true);
+        return true;
+    }
+
+private:
+    bool _getSensorData(robotiq_force_torque_sensor::ft_sensor &msgStream) override
+    {
+        msgStream.Fx = 0.0;
+        msgStream.Fy = 0.0;
+        msgStream.Fz = 0.0;
+        msgStream.Mx = 0.0;
+        msgStream.My = 0.0;
+        msgStream.Mz = 0.0;
+        // Keep publishing rate to 100Hz
+        ros::Duration(0.01).sleep();
```

Comment:
ros::Rate will adjust its sleep time to try to make the rate actually 100Hz rather than 100+Hz you'd see with ros::Duration

```suggestion
        ros::Rate(100).sleep();
```

Reply:
D'oh! Temporaries won't work!

Reply:
(D'oh)^2 "100+Hz you'd see with ros::Duration" -> Since time is in the denominator it'll be less than 100Hz (like ~98) :stuck_out_tongue_winking_eye: